Firebase.CloudFirestore/source/Firebase.CloudFirestore/ApiDefinition.cs

```diff
@@ -0,0 +1,689 @@
+		[Export ("documentWithPath:")]
+		DocumentReference GetDocument (string documentPath);
+
+		// -(FIRDocumentReference * _Nonnull)addDocumentWithData:(NSDictionary<NSString *,id> * _Nonnull)data;
+		[Export ("addDocumentWithData:")]
+		DocumentReference AddDocument (NSDictionary<NSString, NSObject> data);
+
+		[Wrap ("AddDocument (data == null ? null : NSDictionary<NSString, NSObject>.FromObjectsAndKeys (System.Linq.Enumerable.ToArray (data.Values), System.Linq.Enumerable.ToArray (data.Keys), data.Keys.Count))", true)]
+		DocumentReference AddDocument (Dictionary<object, object> data);
+
+		// -(FIRDocumentReference * _Nonnull)addDocumentWithData:(NSDictionary<NSString *,id> * _Nonnull)data completion:(void (^ _Nullable)(NSError * _Nullable))completion;
+		[Export ("addDocumentWithData:completion:")]
+		DocumentReference AddDocument (NSDictionary<NSString, NSObject> data, AddDocumentCompletionHandler completion);
+
+		[Wrap ("AddDocument (data == null ? null : NSDictionary<NSString, NSObject>.FromObjectsAndKeys (System.Linq.Enumerable.ToArray (data.Values), System.Linq.Enumerable.ToArray (data.Keys), data.Keys.Count), completion)", true)]
+		DocumentReference AddDocument (Dictionary<object, object> data, AddDocumentCompletionHandler completion);
+	}
+
+	// @interface FIRDocumentChange : NSObject
+	[DisableDefaultCtor]
+	[BaseType (typeof (NSObject), Name = "FIRDocumentChange")]
+	interface DocumentChange
+	{
+		// @property (readonly, nonatomic) FIRDocumentChangeType type;
+		[Export ("type")]
+		DocumentChangeType Type { get; }
+
+		// @property (readonly, nonatomic, strong) FIRDocumentSnapshot * _Nonnull document;
+		[Export ("document", ArgumentSemantic.Strong)]
+		DocumentSnapshot Document { get; }
+
+		// @property (readonly, nonatomic) NSUInteger oldIndex;
+		[Export ("oldIndex")]
+		nuint OldIndex { get; }
+
+		// @property (readonly, nonatomic) NSUInteger newIndex;
+		[Export ("newIndex")]
+		nuint NewIndex { get; }
+	}
+
+	// @interface FIRDocumentListenOptions : NSObject
+	[BaseType (typeof (NSObject), Name = "FIRDocumentListenOptions")]
+	interface DocumentListenOptions
+	{
+		// +(instancetype _Nonnull)options;
+		[Static]
+		[Export ("options")]
+		DocumentListenOptions Create ();
+
+		// @property (readonly, assign, nonatomic) BOOL includeMetadataChanges;
+		[Export ("includeMetadataChanges")]
+		bool IncludeMetadataChanges { get; }
+
+		// -(instancetype _Nonnull)includeMetadataChanges:(BOOL)includeMetadataChanges;
+		[Export ("includeMetadataChanges:")]
+		DocumentListenOptions SetIncludeMetadataChanges (bool includeMetadataChanges);
+	}
+
+	// typedef void (^FIRDocumentSnapshotBlock)(FIRDocumentSnapshot * _Nullable, NSError * _Nullable);
+	delegate void DocumentSnapshotHandler ([NullAllowed] DocumentSnapshot snapshot, [NullAllowed] NSError error);
+
+	// void (^)(NSError * _Nullable)
+	delegate void DocumentActionCompletionHandler ([NullAllowed] NSError error);
```

Comment:
Suggestion: I would suggest replacing instances of this with just `Action<NSError>` there is not really a need for a named delegate here.

Reply:
I like to have named arguments, so when you want to use an anonymous method, you can have an idea of what you have with the argument name instead of just seeing `arg1`. That's the reason I did a delegate.

Reply:
Welp the IDE should help you with that I mean it is typed as `NSError` so there is not much to guess on this one but #YOLO I'm fine if you want to leave it as is

Reply:
I have faced people arguing why we have nameless args in anonymous methods, that they need to hover over the `arg1` to know what type is it haha. So, if this does not affect performance, I think I'll keep with this heh…

Reply:
> I have faced people arguing why we have nameless args in anonymous methods, that they need to hover over the arg1 to know what type is it haha.

🤦‍♂️ 🤦‍♀️ 
![giphy](https://user-images.githubusercontent.com/204671/38826139-8d1ae9c4-4174-11e8-85aa-bca08a7495d5.gif)

Reply:
> I think I'll keep with this heh…

Fine with me :)